### PR TITLE
Update the web API client tutorial for recent language and .NET Core changes.

### LINKED ISSUE #9496: Clarifying using await in Main
"Normally, you would prefer to await the completion of the task, as in the ProcessRepositories method, but that isn't allowed in the Main method." This isn't accurate if you are using C# 7, right? I was able to change my LangVersion property to 7.1 and the signature of the Main method to return a Task to get this to work.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 3bfcc1a8-ddda-810c-17b0-56a506045d67
* Version Independent ID: 5415878b-e60b-5433-7b4b-efc8259908f0
* Content: [Create a REST client using .NET Core](https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/console-webapiclient)
* Content Source: [docs/csharp/tutorials/console-webapiclient.md](https://github.com/dotnet/docs/blob/master/docs/csharp/tutorials/console-webapiclient.md)
* Product: **dotnet-csharp**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### LINKED ISSUE #16162: Difficult to follow. No clear sequence of steps.
Pretty poor for Microsoft. Dives in at the low level, yet fails to give us a clear sequence of steps that will get this thing up and running. I've been more than an hour trying to find out what this project is supposed to be called, because you don't mention it, and I can't find it in the samples. Very, very poor. Come on, Microsoft. Buck up!

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 3bfcc1a8-ddda-810c-17b0-56a506045d67
* Version Independent ID: 5415878b-e60b-5433-7b4b-efc8259908f0
* Content: [Create a REST client using .NET Core](https://docs.microsoft.com/en-us/dotnet/csharp/tutorials/console-webapiclient#feedback)
* Content Source: [docs/csharp/tutorials/console-webapiclient.md](https://github.com/dotnet/docs/blob/master/docs/csharp/tutorials/console-webapiclient.md)
* Product: **dotnet-csharp**
* Technology: **csharp-fundamentals**
* GitHub Login: @BillWagner
* Microsoft Alias: **wiwagn**

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes dotnet/docs#9496
Fixes dotnet/docs#16162

Update to .NET Core 3.1, using async Main, and System.Text.Json for deserialization.

Also, clarify the project name in the create project step.

The associated code changes are in dotnet/samples#1903, but that PR is independent of this one.

### CODE REVIEW

File: docs/csharp/tutorials/console-webapiclient.md [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ---
 title: Create a REST client using .NET Core
 description: This tutorial teaches you a number of features in .NET Core and the C# language.
-ms.date: 03/06/2017
+ms.date: 01/09/2019
```
Comment (gewarren) [inline docs/csharp/tutorials/console-webapiclient.md]:
```suggestion
ms.date: 01/09/2020
```

File: docs/csharp/tutorials/console-webapiclient.md [hunk 2]
```diff
@@ -42,8 +42,13 @@ comfortable with.
 
 The first step is to create a new application. Open a command prompt and
 create a new directory for your application. Make that the current
-directory. Type the command `dotnet new console` at the command prompt. This
-creates the starter files for a basic "Hello World" application. As this is a
+directory. Type the following command in a console window:
```
Comment (gewarren) [inline docs/csharp/tutorials/console-webapiclient.md]:
```suggestion
directory. Enter the following command in a console window:
```

for accessibility...

File: docs/csharp/tutorials/console-webapiclient.md [hunk 3]
```diff
@@ -63,29 +68,12 @@ needs additional libraries for some of its features, you add those
 dependencies into your C# project (\*.csproj) file. For our example, you'll need to add the `System.Runtime.Serialization.Json` package
```
Comment (gewarren) [inline docs/csharp/tutorials/console-webapiclient.md]:
```suggestion
dependencies into your C# project (\*.csproj) file. For our example, you'll need to add the `System.Runtime.Serialization.Json` package,
```

File: docs/csharp/tutorials/console-webapiclient.md [hunk 4]
```diff
@@ -126,13 +114,12 @@ Next, rename the namespace defined in the `namespace` statement from its default
 
 Next, update the `Main` method to call this method. The
 `ProcessRepositories` method returns a Task, and you shouldn't exit the
-program before that task finishes. Therefore, you must use the `Wait`
-method to block and wait for the task to finish:
+program before that task finishes. Therefore, you must change the signature of `Main`. Add the `async` modifier, and change the return type to `Task`. Then, in the body of the method, add a call to `ProcessRepositories`. Make sure that you `await` that method call:
```
Comment (gewarren) [inline docs/csharp/tutorials/console-webapiclient.md]: "Make sure that you `await` that method call" -> I'm not sure how well that will translate into other languages while keeping await unlocalized.

File: docs/csharp/tutorials/console-webapiclient.md [hunk 5]
```diff
@@ -203,18 +190,16 @@ At this point, you've written the code to retrieve a response from a web server,
 the text that is contained in that response. Next, let's convert that JSON response into C#
 objects.
 
-The JSON Serializer converts JSON data into C# Objects. Your first task is to define a C# class
-type to contain the information you use from this response. Let's build this slowly, so start with
-a simple C# type that contains the name of the repository:
+The <xref:System.Text.Json.JsonSerializer?displayProperty=nameWithType> class serializes objects to JSOn and deserializes JSON into objects. Start by defining a class to represent the `repo` JSON object returned from the GitHub API:
```
Comment (gewarren) [inline docs/csharp/tutorials/console-webapiclient.md]:
```suggestion
The <xref:System.Text.Json.JsonSerializer?displayProperty=nameWithType> class serializes objects to JSON and deserializes JSON into objects. Start by defining a class to represent the `repo` JSON object returned from the GitHub API:
```

File: docs/csharp/tutorials/console-webapiclient.md [hunk 6]
```diff
@@ -228,40 +213,28 @@ The JSON serializer will ignore information that is not included in the class ty
...
-using System.Collections.Generic;
-using System.Runtime.Serialization.Json;
+var streamTask = client.GetStreamAsync("https://api.github.com/orgs/dotnet/repos");
+var repositories = await JsonSerializer.DeserializeAsync<List<Repository>>(await streamTask);
+return repositories;
 ```
 
-Next, you'll use the serializer to convert JSON into C# objects. Replace the call to
-<xref:System.Net.Http.HttpClient.GetStringAsync(System.String)> in your `ProcessRepositories` method with the following two lines:
+You're using a new namespace, so you'll need to add it as well at the top of the file:
```
Comment (gewarren) [inline docs/csharp/tutorials/console-webapiclient.md]:
```suggestion
You're using a new namespace, so you'll need to add it at the top of the file as well:
```

File: docs/csharp/tutorials/console-webapiclient.md [hunk 7]
```diff
@@ -228,40 +213,28 @@ The JSON serializer will ignore information that is not included in the class ty
...
-The declaration of <xref:System.Runtime.Serialization.Json.DataContractJsonSerializer.ReadObject(System.IO.Stream)> declares that it returns an object of type <xref:System.Object?displayProperty=nameWithType>. <xref:System.Runtime.Serialization.Json.DataContractJsonSerializer.ReadObject(System.IO.Stream)> will return the type you specified when you constructed it (`List<repo>` in
-this tutorial). If the conversion does not succeed, the `as` operator evaluates to `null`,
-instead of throwing an exception.
+up to now, you've only seen them as part of an assignment statement. The `Deserialize` method is *generic*, which means you must supply type arguments for what kind of objects should be created from the JSON text. In this example, you're deserializing to a `List<Repository>`, which is another generic object, the <xref:System.Collections.Generic.List%601?displayProperty=nameWithType>. The `List<>` class stores a collection of objects. The type argument declares the type of objects stored in the `List<>`. The Json text represents a collection of `repo` objects, so the type argument is a `Repository`.
```
Comment (gewarren) [inline docs/csharp/tutorials/console-webapiclient.md]:
```suggestion
up to now, you've only seen them as part of an assignment statement. The `Deserialize` method is *generic*, which means you must supply type arguments for what kind of objects should be created from the JSON text. In this example, you're deserializing to a `List<Repository>`, which is another generic object, the <xref:System.Collections.Generic.List%601?displayProperty=nameWithType>. The `List<>` class stores a collection of objects. The type argument declares the type of objects stored in the `List<>`. The Json text represents a collection of repo objects, so the type argument is `Repository`.
```

Comment (gewarren) [reply docs/csharp/tutorials/console-webapiclient.md]: Changes are in the last sentence only.

File: docs/csharp/tutorials/console-webapiclient.md [hunk 8]
```diff
@@ -341,30 +270,8 @@ This change means you need to change the code that writes the name of each repos
...
-
-```csharp
-public string Name
-{
-    get { return this._name; }
-    set { this._name = value; }
-}
-private string _name;
-```
+Do a `dotnet run` to make sure you've got the mappings correct. You should
```
Comment (gewarren) [inline docs/csharp/tutorials/console-webapiclient.md]:
```suggestion
Execute `dotnet run` to make sure you've got the mappings correct. You should
```

File: docs/csharp/tutorials/console-webapiclient.md [hunk 9]
```diff
@@ -453,32 +356,18 @@ this fashion in the JSON response:
 
 That format does not follow any of the standard .NET <xref:System.DateTime> formats. Because of that, you'll need to write
 a custom conversion method. You also probably don't want the raw string exposed to users of the `Repository`
-class. Attributes can help control that as well. First, define a `private` property that will hold the
-string representation of the date time in your `Repository` class:
+class. Attributes can help control that as well. First, define a `public` property that will hold the
+string representation of the date time in your `Repository` class and a `LastPush` readonly property that returns a formatted string representing the date returned:
```
Comment (gewarren) [inline docs/csharp/tutorials/console-webapiclient.md]:
```suggestion
string representation of the date and time in your `Repository` class and a `LastPush` `readonly` property that returns a formatted string that represents the returned date:
```

File: docs/csharp/tutorials/console-webapiclient.md [hunk 10]
```diff
@@ -453,32 +356,18 @@ this fashion in the JSON response:
...
-    }
-}
+public DateTime LastPush =>
+    DateTime.ParseExact(JsonDate, "yyyy-MM-ddTHH:mm:ssZ", CultureInfo.InvariantCulture);
 ```
 
-Let's go over the new constructs above. The `IgnoreDataMember` attribute instructs the serializer
-that this type should not be read to or written from any JSON object. This property contains only a
-`get` accessor. There is no `set` accessor. That's how you define a *read-only* property in C#. (Yes,
+Let's go over the new constructs above. The `LastPush` property is defined using an *expression bodied member* for hte `get` accessor. There is no `set` accessor. Omitting the `set` accessor is how you define a *read-only* property in C#. (Yes,
```
Comment (gewarren) [inline docs/csharp/tutorials/console-webapiclient.md]:
```suggestion
Let's go over the new constructs we just defined. The `LastPush` property is defined using an *expression-bodied member* for the `get` accessor. There is no `set` accessor. Omitting the `set` accessor is how you define a *read-only* property in C#. (Yes,
```

**Review by gewarren — APPROVED** [review]
LG, left some suggestions.

File: docs/csharp/tutorials/console-webapiclient.md (hunk 1 above)
Comment (BillWagner) [reply docs/csharp/tutorials/console-webapiclient.md]: Doh!

**Pull request merged**